ocw/lib/openstack.py

```diff
@@ -0,0 +1,108 @@
+from datetime import datetime, timezone
+from typing import Dict
+from dateutil.parser import parse
+import openstack
+from openstack.exceptions import OpenStackCloudException
+from webui.PCWConfig import PCWConfig
+from .provider import Provider
+
+
+class Openstack(Provider):
+    __instances: Dict[str, "Openstack"] = {}
+
+    def __init__(self, namespace: str):
+        super().__init__(namespace)
+        self.client()
+
+    def __new__(cls, vault_namespace: str):
+        if vault_namespace not in Openstack.__instances:
+            Openstack.__instances[vault_namespace] = self = object.__new__(cls)
+            self.__client = None
+        return Openstack.__instances[vault_namespace]
+
+    def client(self) -> None:
+        if self.__client is None:
+            self.__client = openstack.connect(
+                insecure=True,  # Trust the certificate
+                auth_url=self.get_data('auth_url'),
+                project_name=self.get_data('project_name'),
+                username=self.get_data('username'),
+                password=self.get_data('password'),
+                region_name=self.get_data('region_name'),
+                user_domain_name=self.get_data('user_domain_name'),
+                project_domain_name=self.get_data('project_domain_id'),
+            )
+        return self.__client
+
+    def cleanup_all(self) -> None:
+        self._cleanup_instances()
+        self._cleanup_images()
+        self._cleanup_keypairs()
```

Comment:
while I do see 100% code coverage as our goal there are some lines which more important to cover and there are others which just "nice to have " . this function is second one , because only thing what this function is doing is calling other functions so only test which we can do "does it call this 3 functions ?" which IMHO just "nice to have"